runtime/src/comm/ofi/comm-ofi.c

```diff
@@ -2099,11 +2107,51 @@ void init_ofiForAms(void) {
 }
 
 
-#if 0
+static inline void amRequestNop(c_nodeid_t, chpl_bool, struct perTxCtxInfo_t*);
+
 static
-void init_ofiPerThread(void) {
+void init_ofiConnections(void) {
+  //
+  // With providers that do dynamic endpoint connection we have seen
+  // fairly dramatic connection overheads under certain circumstances.
+  // As an aid to performance analysis, here we allow for forcing the
+  // endpoint connections to be established early, during startup,
+  // rather than later during timed sections of user code.
+  //
+  if (!providerInUse(provType_tcp)
+      && !providerInUse(provType_verbs)) {
+    return;
+  }
+
+  if (!chpl_env_rt_get_bool("COMM_OFI_CONNECT_EAGERLY", false)) {
+    return;
+  }
+
+  //
+  // We do this by firing a no-op AM to every remote node from every
+  // still-inactive tx context.  (In effect, this means from all tx
+  // contexts that aren't already bound to AM handlers.)  At present we
+  // use blocking AMs for this, but if needed we could delay blocking
+  // until after the last one had been initiated.  One way or another,
+  // we need to not return until after all the connections have been
+  // established.
+  //
+  for (int i = 0; i < tciTabLen; i++) {
+    struct perTxCtxInfo_t* tcip = &tciTab[i];
+    if (!tcip->bound) {
+      CHK_TRUE(tciAllocTabEntry(tcip));
+      for (c_nodeid_t node = 0; node < chpl_numNodes; node++) {
```

Comment:
I wonder if we should offset/stagger the node order. Right now all nodes will in lockstep connect to node 0, then 1. I wonder if we should offset by nodeid.

Reply:
I did some quick experiments on Osprey, and I did not see any speed improvement from staggering, so may not be worthwhile, but curious to see what you think.

Reply:
Ironically, I thought about doing this and then decided I might get dinged for being too fancy. I suspect it's something that would only help at scale, but it is more elegant and I'm happy to give it a try.

Reply:
We have definitely seen scaling issues in the past from all nodes hitting specific nodes in lockstep, so I'd be in favor of this change even if we don't see an impact at small scale. It might not matter if we switch to non-blocking AMs though, so up to you if you want to make the change.